### PR TITLE
client: add function useCurrentServer instead of "r == source"

### PR DESCRIPTION
**Pull request by 745150400**
Signed-off-by: 745150400 <745150400@qq.com>

### CODE REVIEW

**Comment by tomponline** [comment]
What was the aim behind this change?

**Comment by 745150400** [comment]
> 



> What was the aim behind this change?

When I use the lxd client in my own project, the incoming server is actually the same server many times, but it is not judged, so I wonder if the judgment can be made more accurate

**Comment by stgraber** [comment]
If going down that path, the function should probably be comparing the server certificates instead of trying to match based on addresses. It also should retain a short path at the beginning of the function checking if source and destination are the same struct.

**Comment by tomponline** [comment]
Also we should need to compare the state of each client, as one may be setup to use different settings (such as project) which make them effectively not the same.

File: client/lxd_instances.go [hunk 1]
```diff
@@ -442,13 +442,40 @@ func (r *ProtocolLXD) tryCreateInstance(req api.InstancesPost, urls []string, op
...
+
+	info, err := r.GetConnectionInfo()
+	if err != nil {
+		return false, err
+	}
+
+	return serverInfo.Protocol == info.Protocol &&
+		serverInfo.Certificate == info.Certificate &&
+		serverInfo.Project == info.Project, nil
+}
```
Comment (stgraber) [inline client/lxd_instances.go:466]: I think it also needs to consider `clusterTarget` which currently isn't exposed in `ConnectionInfo`.

So I think we'd want to add a `Target string` to `ConnectionInfo`, expose it through `GetConnectionInfo` and then compare it in here.

File: client/lxd_instances.go [hunk 2]
```diff
@@ -442,13 +442,40 @@ func (r *ProtocolLXD) tryCreateInstance(req api.InstancesPost, urls []string, op
 	return &rop, nil
 }
 
+func (r *ProtocolLXD) isCurrentServer(server Server) (bool, error) {
+	if r == server {
+		return true, nil
+	}
+	if server == nil {
+		return false, fmt.Errorf("server is null")
```
Comment (stgraber) [inline client/lxd_instances.go:450]: Should be `Provided server is nil` probably.

File: client/lxd_instances.go [hunk 3]
```diff
@@ -442,13 +442,40 @@ func (r *ProtocolLXD) tryCreateInstance(req api.InstancesPost, urls []string, op
 	return &rop, nil
 }
 
+func (r *ProtocolLXD) isCurrentServer(server Server) (bool, error) {
+	if r == server {
+		return true, nil
+	}
```
Comment (stgraber) [inline client/lxd_instances.go:448]: Line break here

File: client/lxd_instances.go [hunk 4]
```diff
@@ -442,13 +442,40 @@ func (r *ProtocolLXD) tryCreateInstance(req api.InstancesPost, urls []string, op
 	return &rop, nil
 }
 
+func (r *ProtocolLXD) isCurrentServer(server Server) (bool, error) {
```
Comment (stgraber) [inline client/lxd_instances.go:445]: Can we make the function just return a `bool`, no `error` and return `false` if it gets an error from GetConnectionInfo?

Currently `GetConnectionInfo` can't return an error anyway, so this should be fine and will make it much easier to use (can go straight in the if statement).

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
I've done a few edits to deal with the test failures and cleaning things up a bit.
Basically:
 - Move the function over to lxd.go (right around GetConnectionInfo)
 - Renamed to isSameServer
 - Added comments
 - Renamed info structs to srcInfo and dstInfo
 - Made it return false if source or remote is nil
 - Make it return false if local struct has un-initialized Server struct

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**